### PR TITLE
API Add new BackupCode auth method and refactor LoginInterface slightly

### LINKED ISSUE #4: Add backup code method - backend
See https://github.com/Firesphere/silverstripe-bootstrapmfa/blob/master/src/Models/BackupCode.php

This will need a DataObject and an implementation of the MethodInterface etc interfaces.

Backup codes will be generated for a user (Member has_many BackupCodes) when their first MFA method is registered, or when they've asked for them to be reset via the admin area.

Designs:
* https://projects.invisionapp.com/share/3PNSKZQYBJZ#/screens/322766388
* https://projects.invisionapp.com/share/3PNSKZQYBJZ#/screens/321186218

### PR DESCRIPTION
**Pull request by ScopeyNZ**
- LoginInterface now has a RegisteredMethod as an argument to both "start" and "verify"
- Removes HandlerInterface as the "start" method now has different arguments for Register and Login
- Adds the Method, LoginHandler and RegistrationHandler for backup codes. Implemented using PHPs password_hash API

Resolves #4

### CODE REVIEW

File: src/BackupCode/Method.php [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+    }
+
+    /**
+     * Return the LoginHandler that is used to start and verify login attempts with this method
+     *
+     * @return LoginHandlerInterface
+     */
+    public function getLoginHandler()
+    {
+        return new LoginHandler();
```
Comment (robbieaverill) [inline src/BackupCode/Method.php]: Can we use Injector for this please?

```suggestion
        return Injector::inst()->create(LoginHandler::class);
```

BYO namespace

File: src/BackupCode/RegisterHandler.php [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+    }
+
+    /**
+     * Provide a localised URL to a support article about the registration process for this MFA Method.
+     *
+     * @return string
+     */
+    public function getSupportLink()
+    {
+        return '';
```
Comment (robbieaverill) [inline src/BackupCode/RegisterHandler.php]: maybe add a todo here, I assume it'll come later

File: tests/php/BackupCode/LoginHandlerTest.php [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+        list ($request, $store, $method) = $this->scaffoldVerifyParams('');
+        $this->assertFalse($handler->verify($request, $store, $method), 'Empty codes are handled');
+
+        // Test a code param is not provided (null value)
+        list ($request, $store, $method) = $this->scaffoldVerifyParams(null);
+        $this->assertFalse($handler->verify($request, $store, $method), 'Null input is handled');
+
+        // Test a code that's way too long
+        list ($request, $store, $method) = $this->scaffoldVerifyParams(str_pad('', 10000, 'code'));
+        $this->assertFalse($handler->verify($request, $store, $method), 'Long codes are handled');
```
Comment (robbieaverill) [inline tests/php/BackupCode/LoginHandlerTest.php]: These look like they could be different cases with a dataProvider

File: tests/php/BackupCode/LoginHandlerTest.php [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+    }
+
+    public function testVerifyReturnsTrueOnValidCode()
+    {
+        $handler = new LoginHandler();
+
+        // Test a code with invalid characters
+        list ($request, $store, $method) = $this->scaffoldVerifyParams('123456');
+        $this->assertTrue($handler->verify($request, $store, $method));
```
Comment (robbieaverill) [inline tests/php/BackupCode/LoginHandlerTest.php]: Mix this into the previous one with a data provider maybe?

**Comment by ScopeyNZ** [comment]
Thanks @robbieaverill I've addressed your feedback 🙂

**Pull request merged**